### PR TITLE
SCRUM-4436 improve construct association load performance

### PR DESCRIPTION
**Pull request by cmpich**

### CODE REVIEW

File: src/main/java/org/alliancegenome/curation_api/dao/ConstructDAO.java [hunk 1]
```diff
@@ -1,15 +1,42 @@
...
 
+	public Map<String, Long> getConstructIdMap() {
+		if (constructIdMap.size() > 0) {
+			return constructIdMap;
+		}
+		Query q = entityManager.createNativeQuery("SELECT a.id, a.modEntityId, a.modInternalId FROM Reagent as a where exists (select * from construct as g where g.id = a.id)");
+		List<Object[]> ids = q.getResultList();
+		ids.forEach(record -> {
+			if (record[1] != null) {
+				constructIdMap.put((String) record[1], (long)record[0]);
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/dao/ConstructDAO.java:28]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'typecast' is not followed by whitespace.

File: src/main/java/org/alliancegenome/curation_api/dao/ConstructDAO.java [hunk 2]
```diff
@@ -1,15 +1,42 @@
...
+			}
+			if (record[2] != null) {
+				constructIdMap.put((String) record[2], (long)record[0]);
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/dao/ConstructDAO.java:31]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'typecast' is not followed by whitespace.

File: src/main/java/org/alliancegenome/curation_api/dao/GenomicEntityDAO.java [hunk 3]
```diff
@@ -1,15 +1,43 @@
...
 
+	public Map<String, Long> getGenomicEntityIdMap() {
+		if (genomicEntityIdMap.size() > 0) {
+			return genomicEntityIdMap;
+		}
+		Query q = entityManager.createNativeQuery("SELECT a.id, a.modEntityId, a.modInternalId FROM BiologicalEntity as a where exists (select * from genomicentity as g where g.id = a.id)");
+		List<Object[]> ids = q.getResultList();
+		ids.forEach(record -> {
+			if (record[1] != null) {
+				genomicEntityIdMap.put((String) record[1], (long)record[0]);
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/dao/GenomicEntityDAO.java:28]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'typecast' is not followed by whitespace.

File: src/main/java/org/alliancegenome/curation_api/dao/GenomicEntityDAO.java [hunk 4]
```diff
@@ -1,15 +1,43 @@
...
+			}
+			if (record[2] != null) {
+				genomicEntityIdMap.put((String) record[2], (long)record[0]);
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/dao/GenomicEntityDAO.java:31]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'typecast' is not followed by whitespace.

**Review by oblodgett — CHANGES_REQUESTED** [review]
The problem with this cache is that it gets set once for the life of the running application. So any updates will not get reflected in the map.

**Comment by cmpich** [comment]
The cache only involves constructs, genes, alleles and AGMs. They do not change during the lifetime of this load. And if any of them do change in case they get updated with a load running in parallel then yes. But this load could also run out of sync without the caching. We have a general order of loads: Gene, allele, AGM then the annotations...

**Comment by oblodgett** [comment]
yes but the cache is under a `@ApplicationScoped` which means it persists in memory for the duration of the server running. So in the case of production which is up for weeks. Then this cache will be out of date after the first time the load is run.

**Comment by oblodgett** [comment]
If the cache was under a `RequestScoped` then yes it would be gone once the request is finished. Other items we do cache under ApplicationScoped (Organization, Ontology Terms, Dataproviders, etc) as those will almost never change. But we specially don't cache Mod objects for this very reason.

**Comment by cmpich** [comment]
Changed the scope to ensure that the cache is not tied to the application running but only for the individual load run.

**Review by oblodgett — CHANGES_REQUESTED** [review]
Changing the scope of the DAO is not the answer. Because now every request to the DAO creates a new connection to the database. The solution should be moving the cache to the server.

File: src/main/java/org/alliancegenome/curation_api/dao/ConstructDAO.java [hunk 5]
```diff
@@ -1,15 +1,35 @@
 package org.alliancegenome.curation_api.dao;
 
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.context.RequestScoped;
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/dao/ConstructDAO.java:4]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unused import - jakarta.enterprise.context.RequestScoped.

File: src/main/java/org/alliancegenome/curation_api/dao/GenomicEntityDAO.java [hunk 6]
```diff
@@ -1,9 +1,14 @@
 package org.alliancegenome.curation_api.dao;
 
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.context.RequestScoped;
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/dao/GenomicEntityDAO.java:4]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unused import - jakarta.enterprise.context.RequestScoped.

File: src/main/java/org/alliancegenome/curation_api/services/GenomicEntityService.java [hunk 7]
```diff
@@ -1,5 +1,6 @@
 package org.alliancegenome.curation_api.services;
 
+import jakarta.persistence.Query;
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/services/GenomicEntityService.java:3]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unused import - jakarta.persistence.Query.

File: src/main/java/org/alliancegenome/curation_api/services/GenomicEntityService.java [hunk 8]
```diff
@@ -11,6 +12,10 @@
 import jakarta.enterprise.context.RequestScoped;
 import jakarta.inject.Inject;
 
+import java.util.HashMap;
+import java.util.List;
```
Comment (devopsagr) [inline src/main/java/org/alliancegenome/curation_api/services/GenomicEntityService.java:16]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Unused import - java.util.List.

**Review by oblodgett — APPROVED** [review]

**Pull request merged**